File: sdk/python/src/syfthub_sdk/_pagination.py
"""Pagination utilities for SyftHub SDK."""

from __future__ import annotations

from collections.abc import Callable, Iterator
from typing import Generic, TypeVar

from pydantic import BaseModel

T = TypeVar("T", bound=BaseModel)

# Type alias for fetch functions
FetchFn = Callable[[int, int], list[dict[str, object]]]


class PageIterator(Generic[T]):
    """Lazy pagination iterator that fetches pages on demand.

    Example usage:
        # Iterate through all items
        for endpoint in client.hub.browse():
            print(endpoint.name)

        # Get just the first page
        first_page = client.hub.browse().first_page()

        # Get all items as a list
        all_items = client.hub.browse().all()

        # Get first 50 items
        items = client.hub.browse().take(50)
    """

    def __init__(
        self,
        fetch_fn: FetchFn,
        model_class: type[T],
        page_size: int = 20,
    ) -> None:
        """Initialize the page iterator.

        Args:
            fetch_fn: Function that takes (skip, limit) and returns list of dicts
            model_class: Pydantic model class to parse items into
            page_size: Number of items per page (default 20)
        """
        self._fetch_fn = fetch_fn
        self._model_class = model_class
        self._page_size = page_size
        self._reset()

    def _reset(self) -> None:
        """Reset iterator state for fresh iteration."""
        self._buffer: list[T] = []
        self._current_page = 0
        self._exhausted = False
        self._started = False

    def _fetch_page(self, page: int) -> list[T]:
        """Fetch a single page and convert to models."""
        skip = page * self._page_size
        raw_items = self._fetch_fn(skip, self._page_size)

        # Convert dicts to model instances
        items = [self._model_class.model_validate(item) for item in raw_items]

        # Check if this is the last page
        if len(items) < self._page_size:
            self._exhausted = True

        return items

    def __iter__(self) -> Iterator[T]:
        """Return iterator (resets state for fresh iteration)."""
        self._reset()
        return self

    def __next__(self) -> T:
        """Return next item, fetching pages as needed."""
        # If we have items in buffer, return next one
        if self._buffer:
            return self._buffer.pop(0)

        # If exhausted, stop iteration
        if self._exhausted:
            raise StopIteration

        # Fetch next page
        self._started = True
        page_items = self._fetch_page(self._current_page)
        self._current_page += 1

        if not page_items:
            self._exhausted = True
            raise StopIteration

        # Buffer all but first item
        self._buffer = page_items[1:]
        return page_items[0]

    def first_page(self) -> list[T]:
        """Get just the first page of results.

        Returns:
            List of items from the first page
        """
        return self._fetch_page(0)

    def all(self) -> list[T]:
        """Fetch all pages and return as a single list.

        Warning: This loads all items into memory.

        Returns:
            List of all items across all pages
        """
        return list(self)

    def take(self, n: int) -> list[T]:
        """Get the first n items (may span multiple pages).

        Args:
            n: Maximum number of items to return

        Returns:
            List of up to n items
        """
        result: list[T] = []
        for item in self:
            result.append(item)
            if len(result) >= n:
                break
        return result

File: sdk/golang/syfthub/pagination.go
package syfthub

import (
	"context"
	"encoding/json"
)

// FetchFunc is a function that fetches a page of items.
// It takes context, skip, and limit parameters and returns raw JSON items.
type FetchFunc func(ctx context.Context, skip, limit int) ([]json.RawMessage, error)

// PageIterator provides lazy pagination that fetches pages on demand.
//
// Example usage:
//
//	// Iterate through all items
//	iter := client.Hub.Browse(ctx)
//	for iter.Next(ctx) {
//	    endpoint := iter.Value()
//	    fmt.Println(endpoint.Name)
//	}
//	if err := iter.Err(); err != nil {
//	    log.Fatal(err)
//	}
//
//	// Get just the first page
//	items, err := client.Hub.Browse(ctx).FirstPage(ctx)
//
//	// Get all items as a slice
//	items, err := client.Hub.Browse(ctx).All(ctx)
//
//	// Get first 50 items
//	items, err := client.Hub.Browse(ctx).Take(ctx, 50)
type PageIterator[T any] struct {
	fetchFn  FetchFunc
	pageSize int

	// Current state
	buffer      []T
	currentPage int
	exhausted   bool
	err         error
	current     T
}

// NewPageIterator creates a new page iterator.
func NewPageIterator[T any](fetchFn FetchFunc, pageSize int) *PageIterator[T] {
	if pageSize <= 0 {
		pageSize = 20
	}
	return &PageIterator[T]{
		fetchFn:  fetchFn,
		pageSize: pageSize,
	}
}

// Reset resets the iterator state for fresh iteration.
func (p *PageIterator[T]) Reset() {
	p.buffer = nil
	p.currentPage = 0
	p.exhausted = false
	p.err = nil
	var zero T
	p.current = zero
}

// fetchPage fetches a single page and converts to typed items.
func (p *PageIterator[T]) fetchPage(ctx context.Context, page int) ([]T, error) {
	skip := page * p.pageSize
	rawItems, err := p.fetchFn(ctx, skip, p.pageSize)
	if err != nil {
		return nil, err
	}

	// Convert raw JSON to typed items
	items := make([]T, 0, len(rawItems))
	for _, raw := range rawItems {
		var item T
		if err := json.Unmarshal(raw, &item); err != nil {
			return nil, err
		}
		items = append(items, item)
	}

	// Check if this is the last page
	if len(items) < p.pageSize {
		p.exhausted = true
	}

	return items, nil
}

// Next advances the iterator to the next item.
// Returns true if there is a next item, false otherwise.
// Call Value() to get the current item after Next() returns true.
func (p *PageIterator[T]) Next(ctx context.Context) bool {
	// If we have items in buffer, use the next one
	if len(p.buffer) > 0 {
		p.current = p.buffer[0]
		p.buffer = p.buffer[1:]
		return true
	}

	// If exhausted, stop iteration
	if p.exhausted {
		return false
	}

	// Fetch next page
	pageItems, err := p.fetchPage(ctx, p.currentPage)
	if err != nil {
		p.err = err
		return false
	}
	p.currentPage++

	if len(pageItems) == 0 {
		p.exhausted = true
		return false
	}

	// Set current to first item, buffer the rest
	p.current = pageItems[0]
	if len(pageItems) > 1 {
		p.buffer = pageItems[1:]
	}

	return true
}

// Value returns the current item.
// Should only be called after Next() returns true.
func (p *PageIterator[T]) Value() T {
	return p.current
}

// Err returns any error that occurred during iteration.
func (p *PageIterator[T]) Err() error {
	return p.err
}

// FirstPage returns just the first page of results.
func (p *PageIterator[T]) FirstPage(ctx context.Context) ([]T, error) {
	return p.fetchPage(ctx, 0)
}

// All fetches all pages and returns as a single slice.
// Warning: This loads all items into memory.
func (p *PageIterator[T]) All(ctx context.Context) ([]T, error) {
	p.Reset()
	var result []T
	for p.Next(ctx) {
		result = append(result, p.Value())
	}
	if err := p.Err(); err != nil {
		return nil, err
	}
	return result, nil
}

// Take returns the first n items (may span multiple pages).
func (p *PageIterator[T]) Take(ctx context.Context, n int) ([]T, error) {
	p.Reset()
	result := make([]T, 0, n)
	for p.Next(ctx) {
		result = append(result, p.Value())
		if len(result) >= n {
			break
		}
	}
	if err := p.Err(); err != nil {
		return nil, err
	}
	return result, nil
}

// Collect is an alias for All that fetches all pages.
func (p *PageIterator[T]) Collect(ctx context.Context) ([]T, error) {
	return p.All(ctx)
}

// ForEach iterates through all items and calls the callback for each.
func (p *PageIterator[T]) ForEach(ctx context.Context, fn func(T) error) error {
	p.Reset()
	for p.Next(ctx) {
		if err := fn(p.Value()); err != nil {
			return err
		}
	}
	return p.Err()
}

// Filter returns a new slice containing only items that match the predicate.
func (p *PageIterator[T]) Filter(ctx context.Context, predicate func(T) bool) ([]T, error) {
	p.Reset()
	var result []T
	for p.Next(ctx) {
		if predicate(p.Value()) {
			result = append(result, p.Value())
		}
	}
	if err := p.Err(); err != nil {
		return nil, err
	}
	return result, nil
}
